pkg/cmd/docs-issue-generation/github.go

```diff
@@ -0,0 +1,294 @@
+			}
+		}
+		// runs if there are more than 100 results
+		if x.Commits.PageInfo.HasNextPage {
+			additionalCommits, err := searchCockroachPRCommits(x.Number, x.Commits.PageInfo.EndCursor, prCommitsToExclude)
+			if err != nil {
+				return false, "", nil, err
+			}
+			commits = append(commits, additionalCommits...)
+		}
+		if len(commits) > 0 {
+			pr := cockroachPR{
+				Number:      x.Number,
+				Title:       x.Title,
+				Body:        x.Body,
+				BaseRefName: x.BaseRefName,
+				Commits:     commits,
+			}
+			result = append(result, pr)
+		}
+	}
+	pageInfo := search.Data.Search.PageInfo
+	return pageInfo.HasNextPage, pageInfo.EndCursor, result, nil
+}
+
+func searchCockroachPRCommits(
+	pr int, cursor string, prCommitsToExclude map[int]map[string]string,
+) ([]cockroachCommit, error) {
+	hasNextPage, nextCursor, commits, err := searchCockroachPRCommitsSingle(pr, cursor, prCommitsToExclude)
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+	result := commits
+	for hasNextPage {
+		hasNextPage, nextCursor, commits, err = searchCockroachPRCommitsSingle(pr, nextCursor, prCommitsToExclude)
+		if err != nil {
+			fmt.Println(err)
+			return nil, err
+		}
+		result = append(result, commits...)
+	}
+	return result, nil
+}
+
+func searchCockroachPRCommitsSingle(
+	prNumber int, cursor string, prCommitsToExclude map[int]map[string]string,
+) (bool, string, []cockroachCommit, error) {
+	var result []cockroachCommit
+	var search gqlCockroachPRCommit
+	query := `query ($cursor: String, $prNumber: Int!) {
+  repository(owner: "cockroachdb", name: "cockroach") {
+    pullRequest(number: $prNumber) {
+      commits(first: 100, after: $cursor) {
+        edges {
+          node {
+            commit {
+              oid
+              messageHeadline
+              messageBody
+            }
+          }
+        }
+        pageInfo {
+          hasNextPage
+          endCursor
+        }
+      }
+    }
+  }
+}`
+	queryVariables := map[string]interface{}{
+		"prNumber": prNumber,
+	}
+	if cursor != "" {
+		queryVariables["cursor"] = cursor
+	}
+	err := queryGraphQL(
+		query,
+		queryVariables,
+		&search,
+	)
+	if err != nil {
+		fmt.Println(err)
+		return false, "", nil, err
+	}
+	for _, x := range search.Data.Repository.PullRequest.Commits.Edges {
+		matchingDocsIssue := prCommitsToExclude[prNumber][x.Node.Commit.Oid]
+		if nonBugFixRNRE.MatchString(x.Node.Commit.MessageHeadline) && matchingDocsIssue == "" {
+			commit := cockroachCommit{
+				Sha:             x.Node.Commit.Oid,
+				MessageHeadline: x.Node.Commit.MessageHeadline,
+				MessageBody:     x.Node.Commit.MessageHeadline,
+			}
+			result = append(result, commit)
+		}
+	}
+	pageInfo := search.Data.Repository.PullRequest.Commits.PageInfo
+	return pageInfo.HasNextPage, pageInfo.EndCursor, result, nil
+}
+
+var searchCockroachReleaseBranches = func() ([]string, error) {
+	var result []string
+	hasNextPage, nextCursor, branches, err := searchCockroachReleaseBranchesSingle("")
```

Comment:
It looks like we have a lot of small helper functions that are shaped like this: this one, `searchCockroachPRCommits`, `searchCockroachPRs`. It seems you can write the helper function *once*, and use generics to reduce the code duplication. Consider this optional since it's not too much duplicated code.

Reply:
I think that'll be a good exercise for me as a follow-up. I'll need to think about this and strategize a bit.